otri/filtering/stream.py

```diff
@@ -0,0 +1,37 @@
+from typing import Iterable
+
+class Stream(list):
+    '''
+    Collection that uses StreamIter as iterator.
+    '''
+
+    def __iter__(self):
+        return StreamIter(self)
+
+    def insert(self, element):
+        '''
+        Inserts an element at the end of the collection
+        '''
+        super().append(element)
+
+class StreamIter:
+    '''
+    Iterator that removes the items when using them.
+    '''
+
+    def __init__(self, iterable : Iterable):
+        self.iterable = iterable
+
+    def __next__(self):
+        '''
+        Pops the first element of the given collection.
+        '''
+        value = self.iterable[0]
+        del self.iterable[0]
```

Comment:
Funge? Toglierlo lo toglie, ma lo cancella anche dalla memoria penso (???)

Reply:
È quello che vogliamo, vogliamo liberare la memoria dello stream attuale una volta che è stato letto un atomo, se ne occuperà il filtro a ficcare l'atomo in un altro stream per tenerlo vivo (o distruggerlo perché non serve)

Reply:
Il punto è che così distruggi secco l'atomo, il che assume che il ricevente ne abbia fatto una deep copy, è necessario? La garbage collection lo rimuoverebbe ugualmente se non fosse più usato

Reply:
Il garbage collector non lo rimuoverebbe finché la lista non viene più usata, se dobbiamo processare 1TB di atomi e abbiamo tre filtri alla fine ci troveremo con 3TB di stream che contengono i tre stadi di ogni atomo, invece vogliamo che quel 1TB rimanga uno solo

Reply:
Ok provando ho scoperto che un del su un elemento della lista lo rimuove solo, ero convinto che avrebbe cancellato anche l'oggetto. Non lo fa perchè gli elementi di una lista sono oggetti a parte, dei puntatori.